web/source/osk/banner.ts

```diff
@@ -376,94 +482,191 @@ namespace com.keyman.osk {
+      this.recentAccept = false;
+      this.recentRevert = true;
+      this.doUpdate();
     }
 
+    /**
+     * Receives messages from the keyboard that the 'accept' keystroke has been entered.
+     * Should return 'false' if the current state allows accepting a suggestion and act accordingly.
+     * Otherwise, return true.
+     */
+    tryAccept: (source: string) => boolean = function(this: SuggestionManager, source: string): boolean {
+      if(!this.recentAccept && this.selected) {
+        this.doAccept(this.selected);
+        return false;
+      } else if(this.recentAccept && source == 'space') {
+        this.recentAccept = false;
+        return false; // Swallows a single space post-accept.
+      }
+      return true;  // Not yet implemented
+    }.bind(this);
+
+    /**
+     * Receives messages from the keyboard that the 'revert' keystroke has been entered.
+     * Should return 'false' if the current state allows reverting a recently-applied suggestion and act accordingly.
+     * Otherwise, return true.
+     */
+    tryRevert: () => boolean = function(this: SuggestionManager): boolean {
+      if(this.recentAccept) {
+        this.doRevert();
+        return false;
+      } else {
+        return true;
+      }
+    }.bind(this);
+
     /**
      * Function invalidateSuggestions
      * Scope        Public
      * Description  Clears the suggestions in the suggestion banner
      */
-    public invalidateSuggestions: (this: SuggestionBanner) => boolean = 
-        function(this: SuggestionBanner) {
-      this.suggestionList.forEach((option: BannerSuggestion) => {
+    public invalidateSuggestions: (this: SuggestionManager, source: text.prediction.InvalidateSourceEnum) => boolean = 
+        function(this: SuggestionManager, source: string) {
+
+      if(!this.swallowPrediction || source == 'context') {
+        this.recentAccept = false;
+        this.recentRevert = false;
+        this.rejectedSuggestions = [];
+
+        if(source == 'context') {
+          this.swallowPrediction = false;
+          this.initNewContext = true;
+        }
+      }
+
+      this.options.forEach((option: BannerSuggestion) => {
         option.update(null);
       });
     }.bind(this);
 
+    private doUpdate() {
+      let keyman = com.keyman.singleton;
+
+      let suggestions = [];
+      // TODO:  Insert 'current text' if/when valid as the leading option.
+      //        We need the LMLayer to tell us this somehow.
+      if(!this.recentAccept && !this.recentRevert && !this.initNewContext) {
+        // In the meantime, a placeholder:
+        // (generated from the 'true'/original transform)
+        let original = keyman.modelManager.getPredictionState(this.currentTranscriptionID);
+        let nil = new text.TextTransform(original.transform.insert + ' ', original.transform.deleteLeft, original.transform.deleteRight);
+
+        suggestions.push({transform: nil, displayAs: '&lt;keep&gt;', transformId: this.currentTranscriptionID});
```

Comment:
Granted the UX is subject to change.
From the

Reply:
Point 1 - lexical models don't provide us with that ability yet.  That's why I'm using `<keep>` for now.  (It's either "the whole context" or a major language-specific assumption right now.)

Point 2 - That's generally whenever we're willing for said suggestion to actually be automatic - that's something that should probably have a little language model backing as well.  Otherwise, consider the impact on UX for those who believe in double-spacing after a period.  (Hint - that second space would trigger a suggestion!)